tests/integration/shutdown.tcl

```diff
@@ -0,0 +1,157 @@
+                $master set k 1
+                wait_for_ofs_sync $master $replica
+
+                # Pause the replica.
+                exec kill -SIGSTOP $replica_pid
+                after 10
+
+                # Fill upp the replication socket buffers.
+                set junk_size 10000
+                for {set i 0} {$i < $junk_size} {incr i} {
+                    $master set "key.$i.junkjunkjunk" \
+                        "value.$i.blablablablablablablablabla"
+                }
+
+                # Incr k and immediately shutdown master.
+                $master incr k
+                switch $how {
+                    sigterm {
+                        exec kill -SIGTERM $master_pid
+                    }
+                    shutdown {
+                        set rd [redis_deferring_client -1]
+                        $rd shutdown
+                    }
+                }
+
+                # Wake up replica and check if master has waited for it.
+                after 1000
+                exec kill -SIGCONT $replica_pid
+                wait_for_condition 300 1000 {
+                    [$replica get k] eq 2
+                } else {
+                    fail "Master exited before replica could catch up."
+                }
+                assert_equal [expr $junk_size + 1] [$replica dbsize]
+
+                # Check shutdown log messages on master
+                wait_for_log_messages -1 {"*ready to exit, bye bye*"} 0 100 500
+                assert_equal 0 [count_log_message -1 "*Lagging replica*"]
+                verify_log_message -1 "*1 of 1 replicas are in sync*" 0
+            }
+        }
+    } {} {repl external:skip}
+}
+
+test {Shutting down master waits for replica timeout} {
+    start_server {} {
+        start_server {} {
+            set master [srv -1 client]
+            set master_host [srv -1 host]
+            set master_port [srv -1 port]
+            set master_pid [srv -1 pid]
+            set replica [srv 0 client]
+            set replica_pid [srv 0 pid]
+
+            # Config master.
+            $master config set shutdown-timeout 2 ; # seconds
+
+            # Config replica.
+            $replica replicaof $master_host $master_port
+            wait_for_sync $replica
+
+            # Preparation: Set k to 1 on both master and replica.
+            $master set k 1
+            wait_for_ofs_sync $master $replica
+
+            # Pause the replica.
+            exec kill -SIGSTOP $replica_pid
+            after 10
+
+            # Incr k and immediately shutdown master.
+            $master incr k
+            exec kill -SIGTERM $master_pid
+
+            # Wake up replica after master has finished shutting down.
+            after 2500
+            exec kill -SIGCONT $replica_pid
+
+            # Check shutdown log messages on master
+            verify_log_message -1 "*Lagging replica*" 0
+            verify_log_message -1 "*0 of 1 replicas are in sync*" 0
```

Comment:
maybe check that the INCR didn't get to the replica?

Reply:
It only works if we fill the buffers with junk, like in the first test case. Otherwise, the replica will still receive the INCR even after master has shut down, due to flushSlavesOutputBuffers() in finishShutdown. Filling up with junk makes the test very slow in some builds so I was hoping we only need to do that in one of the test cases.

```C
    /* Best effort flush of slave output buffers, so that we hopefully
     * send them pending writes. */
    flushSlavesOutputBuffers();
```

Reply:
you mean we need to write enough data that will not fit into the OS socket buffers?
i guess that's the realistic test we need to write in order to cover the actual problem we're dealing with.
i think we can very quickly fill 10mb and it won't cause any issues.
or even better, keep feeling until introspection shows that there's unwritten data still pending (so it'll work even on a system with huge socket buffers)
let's try.

Reply:
Exactly. Interesting idea. I don't know how to introspect the pending unwritten data though. Is it in INFO?

Reply:
i suppose setting a small `repl-backlog-size`, and watching `mem_total_replication_buffers` is one way.
the other way is do `CLIENT LIST` and look at the `tot-mem` of the replica client.

Reply:
I write junk until `mem_total_replication_buffers` > 1M. Please check if it makes sense.

Reply:
@oranagra The last CI test run failed on this (but the sanitizer test run passed). I don't know how to reliably fill the OS socket send buffer. Do you?

From within Redis it'd be possible to call setsockopt() and getsockopt() with `SO_SNDBUF`. From the test case, perhaps it's possible to use `ss` to inspect the sockets as in [this post](https://stackoverflow.com/a/56279407/975856).

WDYT?

Reply:
i'd rather not.
@yoav-steinberg please take a look at this, you faced similar problems in the client eviction PR.

Reply:
I believe it's not possible to reliably saturate a socket. There is no guarantee that only the send and receive buffers can store data in transit. In a network, data can also sit in routers. Even within the same machine, there kernel might have other internal buffers between the send and receive sockets.

Thus, I suggest that we send a large amount, e.g. 10MB, in the first test cases (the test case where the master waits for the replica). Only 100k is enough on my machine to make it fail without this PR, so 10M is a bit extra. If the feature is broken, this test will fail on at least some machines.

For the negative test case where shutdown times out, I suggest we simply don't try to saturate the socket. We just check the logs on master.

@yoav-steinberg, @oranagra does this make sense?

Reply:
> I believe it's not possible to reliably saturate a socket. 

I don't have any concrete objection to your suggestion, but I don't agree with the reason. 
At some point we'll saturate all the buffers there are, we just need to write code that's robust enough, and doesn't assume anything, eventually the socket will be blocked, and every byte you add will be kept in the client output buffers. 
On some machine 100k will be enough, on others it'll take 10mb, if we add proper exit condition to the loop, we'll be able to trigger what we came to test. 

I'd like to try to avoid constants (times and sizes), and instead add loops that break out when it's right. 

Maybe the filling loops should have a short sleep, and keep writing until it observes that everything it wrote in the last cycle is sitting in the client buffers.

Reply:
Writing 1M at a time, with 10ms sleep between each round, until mem_total_replication_buffers > 2M seems to work.

Reply:
@zuiderkwast I never found a reliable way to saturate the kernel output buffers by assuming a fixed size will work and playing around with linux/socket tweakables. I ended up just writing a look and checking waiting until the client's output buff (or total client memory) reached the size I was aiming for (via `client list`).

Reply:
OK, that's what I tried, using `client list` and `INFO`, but I didn't quite get it right until now. When writing 1M, the repl buffer grows to 1M even when it succeeds writing all of it to the socket. However, it only reaches 2M when the socket is full. Thus, my last attempt seems to work. I hope it's a safe assumption.